### ISSUE TITLE
Same type evaluated differently when in a Tuple or by itself

### ISSUE THREAD

**Issue by 188599** [microsoft/TypeScript#47579]
# Bug Report

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.6.0-dev.20220116#code/C4TwDgpgBAKuEHUCWwAWB7ArsGmwBsIAeGAPigF4oBtGAGigAUBDAJ2CWfxNIF0BuAFChIseMjS4CxMpVhQAPjXpM2HLjwGDBSAHbAIrAGbMAxtACCUAN6Co9qMwBcUXZgC2AI0NCHUTy4AzsCsegDmQgC+2qb4zIGBUFbWUNGCpui6wVBGmLqmAHLowACqgeFShHCiVER2DjBQEAAeBroAJomsEMztmfggUHkA1rroAO661Lz19qQAFAB0y2xhgS7Uy4sp1ADSUHpQwxAg6EawvC7ViCgY2JUye7zkkbwAlJTkKWkZWcA5eVMZQqeCq8DkdT8jRabU6UG6vX6gxGY0m01mUAWW1W6xoWx2+0Ox1O5xglzEkAkqAeJCeL3enxsqW0gly+SKpXKujCD2u82ougg4yS8zeDGsrw+AHopQddEZDIk0NARCBwlAiNQLM95mhwoEAPoABiuqHCRRpFlIHwo5AAbugkO1tGygVyeaCIHyBUKRWKbJKoDK5QrWErUCrwGruRrqBKdXruYaTRSbpJPUQJdbGQ6ndogA)

### 💻 Code

Considering the following code:

```ts
type TypeWithoutTuple<T> = [T, Partial<T>];
type TypeWithTuple<T> = T | [T, Partial<T>];

interface A {
    a: number;
    b: string;
}

class A { }

const funcNotUsingTupleType = <
    T extends readonly unknown[]
    >(...args: [...{ [K in keyof T]: TypeWithoutTuple<T[K]> }]) => { }

const funcUsingTupleType = <
    T extends readonly unknown[]
    >(...args: [...{ [K in keyof T]: TypeWithTuple<T[K]> }]) => { }


funcNotUsingTupleType([new A(), {}]) // infers the typying <[A]>(things_0: ThingNoTuple<A>) => void

funcUsingTupleType([new A(), {}]) // infers the typying <[{}]>(things_0: TypeWithTuple<{}>) => void
```

### 🙁 Actual behavior

The same calls to `funcNotUsingTupleType` and `funcUsingTupleType` have different inferred typyings, even when they are evaluating the same thing.

### 🙂 Expected behavior

The same calls to `funcNotUsingTupleType` and `funcUsingTupleType` to have the same inferred typyings.

**Comment by RyanCavanaugh** [comment]
These are both valid inferences.

In the first case, because `TypeWithoutTuple` is just a tuple type, we can safely collect inference candidates based on the tuple positions.

In the second case, `TypeWithTuple` has the possibility of `T`, so it's not safe to collect inference candidates from the tuple positions, so instead the type of the array literal gets computed (and reduced to `{ }`, since we can't see at this point that `{}` is supposed to be a `Partial<A>`).

This *might* be solved with #30134 but in general an inference onto `type S<T> = T | U<T>` is not going to be able to produce as good as inference as one without it.

**Comment by 188599** [comment]
I see, I figured it might be worth bringing this up here because I had this working with multiple generic types before, something like this:

```ts
const funcUsingTupleTypeWithoutRestParamsGeneric = <T1, T2, T3>(...args: [TypeWithTuple<T1>, TypeWithTuple<T2>?, TypeWithTuple<T3>?]) => { }

funcUsingTupleTypeWithoutRestParamsGeneric([new A(), {}]) // infers the typing <A, unknown, unknown>(args_0: TypeWithTuple<A>, args_1?: unknown, args_2?: unknown) => void
```

And I figured that converting to this other way I mentioned on the issue would work just as well.
